flink-kubernetes-operator/src/main/java/org/apache/flink/kubernetes/operator/observer/Observer.java

```diff
@@ -0,0 +1,162 @@
+    }
+
+    private void observeJmDeployment(
+            FlinkDeployment flinkApp, Context context, Configuration effectiveConfig) {
+        FlinkDeploymentStatus deploymentStatus = flinkApp.getStatus();
+        JobManagerDeploymentStatus previousJmStatus =
+                deploymentStatus.getJobManagerDeploymentStatus();
+
+        if (JobManagerDeploymentStatus.READY == previousJmStatus) {
+            return;
+        }
+
+        if (JobManagerDeploymentStatus.DEPLOYED_NOT_READY == previousJmStatus) {
+            deploymentStatus.setJobManagerDeploymentStatus(JobManagerDeploymentStatus.READY);
+            return;
+        }
+
+        Optional<Deployment> deployment = context.getSecondaryResource(Deployment.class);
+        if (deployment.isPresent()) {
+            DeploymentStatus status = deployment.get().getStatus();
+            DeploymentSpec spec = deployment.get().getSpec();
+            if (status != null
+                    && status.getAvailableReplicas() != null
+                    && spec.getReplicas().intValue() == status.getReplicas()
+                    && spec.getReplicas().intValue() == status.getAvailableReplicas()
+                    && flinkService.isJobManagerPortReady(effectiveConfig)) {
+
+                // typically it takes a few seconds for the REST server to be ready
+                LOG.info(
+                        "JobManager deployment {} in namespace {} port ready, waiting for the REST API...",
+                        flinkApp.getMetadata().getName(),
+                        flinkApp.getMetadata().getNamespace());
+                deploymentStatus.setJobManagerDeploymentStatus(
+                        JobManagerDeploymentStatus.DEPLOYED_NOT_READY);
+                return;
+            }
+            LOG.info(
+                    "JobManager deployment {} in namespace {} exists but not ready yet, status {}",
+                    flinkApp.getMetadata().getName(),
+                    flinkApp.getMetadata().getNamespace(),
+                    status);
+
+            deploymentStatus.setJobManagerDeploymentStatus(JobManagerDeploymentStatus.DEPLOYING);
+            return;
+        }
+
+        deploymentStatus.setJobManagerDeploymentStatus(JobManagerDeploymentStatus.MISSING);
+    }
+
+    private boolean observeFlinkJobStatus(FlinkDeployment flinkApp, Configuration effectiveConfig) {
+
+        // No need to observe job status for session clusters
+        if (flinkApp.getSpec().getJob() == null) {
+            return true;
+        }
+
+        LOG.info("Getting job statuses for {}", flinkApp.getMetadata().getName());
+        FlinkDeploymentStatus flinkAppStatus = flinkApp.getStatus();
+
+        Collection<JobStatusMessage> clusterJobStatuses;
+        try {
+            clusterJobStatuses = flinkService.listJobs(effectiveConfig);
+        } catch (Exception e) {
+            LOG.error("Exception while listing jobs", e);
+            flinkAppStatus.getJobStatus().setState("UNKNOWN");
```

Comment:
I'm confused that whether to set the state to `UNKNOWN` when listing jobs exception. IMO, the listing jobs exception is same as no jobs found, therefore this doesn't update the job status.

Reply:
I have added this so I can simply call observe when we delete the flink resource and only try to cancel if the job is RUNNING.

UNKNOWN does not say that it is running/failing it barely means that the operator cannot access it so flink client operations are not possible. It might not be perfect but I think it makes sense